sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala

```diff
@@ -0,0 +1,213 @@
+ */
+
+package org.apache.spark.sql.rapids
+
+import java.net.URISyntaxException
+
+import ai.rapids.cudf.{ColumnVector, DType, RegexProgram, Scalar, Table}
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.Arm._
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.shims.ShimExpression
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.rapids.shims.RapidsErrorUtils
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+object GpuParseUrl {
+  private val HOST = "HOST"
+  private val PATH = "PATH"
+  private val QUERY = "QUERY"
+  private val REF = "REF"
+  private val PROTOCOL = "PROTOCOL"
+  private val FILE = "FILE"
+  private val AUTHORITY = "AUTHORITY"
+  private val USERINFO = "USERINFO"
+  private val REGEXPREFIX = """(&|^|\?)"""
+  private val REGEXSUBFIX = "=([^&]*)"
+}
+
+case class GpuParseUrl(children: Seq[Expression], 
+    failOnErrorOverride: Boolean = SQLConf.get.ansiEnabled) 
+  extends GpuExpression with ShimExpression with ExpectsInputTypes {
+
+  def this(children: Seq[Expression]) = this(children, SQLConf.get.ansiEnabled)
+
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+  override def prettyName: String = "parse_url"
+
+  import GpuParseUrl._
+  
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      RapidsErrorUtils.parseUrlWrongNumArgs(children.size) match {
+        case res: Some[TypeCheckResult] => return res.get
+        case _ => // error message has been thrown
+      }
+    }
+    super[ExpectsInputTypes].checkInputDataTypes()
+  }
+
+  private def getPattern(key: UTF8String): RegexProgram = {
+    val regex = REGEXPREFIX + key.toString + REGEXSUBFIX
+    new RegexProgram(regex)
+  }
+
+  private def reValid(url: ColumnVector): ColumnVector = {
+    // TODO: Validite the url
+    val regex = """^[^ ]*$"""
+    val prog = new RegexProgram(regex)
+    withResource(url.matchesRe(prog)) { isMatch =>
+      if (failOnErrorOverride) {
+        withResource(isMatch.all()) { allMatch =>
+          if (!allMatch.getBoolean) {
+            val invalidUrl = UTF8String.fromString(url.toString())
+            val exception = new URISyntaxException("", "")
+            throw RapidsErrorUtils.invalidUrlException(invalidUrl, exception)
+          }
+        }
+      }
+      withResource(Scalar.fromNull(DType.STRING)) { nullScalar =>
+        isMatch.ifElse(url, nullScalar)
+      }
+    }
+  }
+
+  private def reMatch(url: ColumnVector, partToExtract: String): ColumnVector = {
+    val regex = """^(([^:/?#]+):)(//((([^:]*:?[^\@]*)\@)?(\[[0-9A-Za-z%.:]*\]|[^/?#:]*)""" + 
```

Comment:
From a performance standpoint capturing all of the groups to turn them into a table, and then only selecting one of them is a real performance problem. You don't have to do it now, but we should at least file a follow on issue to build up the regular expression so we only capture the group that matters.  It would also allow us to document the different parts of the pattern and explain how they are used.

Reply:
Done.